### DISCUSSION TITLE
Don't understand BaseRegion & more

### DISCUSSION THREAD

**Discussion by bmason8** [JaffaKetchup/flutter_map_tile_caching#29] [Q&A] [answered]
Looking for some help with downloading offline maps and in particular how to create a `BaseRegion` for passing into the `DownloadableRegion` constructor.
The 'startDownload' function take a `DownloadableRegion` as it's second parameter and since I get the areas I want downloaded via an api so I don't have the need for the user to draw a box for an area.

So my issue is that when I go to create a `BaseRegion`, it's an abstract class with three function that return either a: 

- `DownloadableRegion`
- `PolygonLayerOptions`
- `List<LatLng>`

So how am I supposed to create a `DownloadableRegion` if one of the parameters is a `BaseRegion` which itself generates a `DownloadableRegion`?

I'm certain this is just me not understanding what's going on but I feel like I'm in a chicken or the egg type situation.
Any help would be appreciated!

**Comment by JaffaKetchup** [comment, marked as answer]
Hi there, and thanks for your question! Below I have written an answer based on what I think your question is: there's a high possibility that I've misunderstood you though, so please don't hesitate to correct me if I'm wrong :)

`BaseRegion` is an abstract class that defines the 'conversion' functions for all regions - you should never need to construct this yourself. `DownloadableRegion` contains a `BaseRegion` and other information needed for downloading. I think you got this part: the part I don't think you got (for reasons I'll explain later) is that you should never need to construct `DownloadableRegion` yourself either.

To resolve your issue, you'll need to use one of `RectangleRegion`, `CircleRegion`(, or `LineRegion` as of v4). All of these extend `BaseRegion`, and represent different shapes.
Then, to construct a `DownloadableRegion`, take your `...Region` and use `.toDownloadable()` on it. You can then pass this to the download functions.

I think the probable reason you got stuck here is because the documentation for this library makes it hard to setup. I am gradually working on this for the next update.
Having said that however, here is another reason): I think you may be using v3. Please upgrade to v4.0.0-dev.6 from pub.dev: it is 99% stable & production-ready, and is unlikely to have more breaking changes.

The major thing about v4 for you though, is that `DownloadableRegion`'s constructor is marked as `@internal`, which would've alerted you to the problem in your editor, thus allowing you to try another way.
`DownloadableRegion` can't be abstract or private because I need to use it internally and elsewhere in the library (not in the same file).

I hope you found this easy enough to understand, and I hope it resolved your issue!

**Reply by bmason8** [reply]
Thanks for the quick reply! I am actually on v4.0.0-dev.6 but apparently was just clueless and didn't notice the @internal tag on `DownloadableRegion`. This makes sense now, thank you for your explanation!

**Reply by JaffaKetchup** [reply]
No problem, glad to help!

Out of interest, which IDE are you using? For me, VS Code gives a 'blue' warning whenever I try to use an internal method. Did the warning not appear for you, or did you just not understand it?

Also just wanted to ask a few questions to smooth off any rough edges for v4, so it would be a huge help for you to spare a few minutes to answer them if you want to and when you have time :)

1. **What is your main use case for this library?**  
For example: Are you building an app with both 'browse caching' and bulk downloading, or just one or the other? Does your app have a dedicated offline mode that you're using this library for?
2. **What is the theme/purpose of your app?**  
For example: Is it a hiking app, a food delivery app or just a general mapping app?
3. **What platform are you developing for?**
For example: Android/iOS/Both
4. **On a scale of 1 (hard) to 10 (easy), how easy was it to setup this library in your app?**  
If possible, explain why you gave it that rating (what was hard, what was easy)
5. **Is there anything you'd like to see added/removed/changed before the Version 4 release?**  
If possible, explain why you chose these things: how would it help you?

Many thanks, and have a great day!

**Reply by bmason8** [reply]
I'm using Android Studio Arctic Fox | 2020.3.1 Patch 3 (which is actually the latest ARM version) on a M1 Pro Macbook Pro.
It gave me a warning but in my haste to figure things out I just ignored it stupidly.

Answers to your questions:

1. Using it for both 'browse caching' and bulk downloading. The app is used by people who will be completely offline at times (network only, it always needs gps) so although it doesn't have a dedicated 'Offline Mode' it requires the ability to view large areas while offline. When the app downloads project data, included in it are areas (rectangles) that the user is likely to work in so I use those for the bulk downloading but also want them to be able to pan over an area so it gets added to the cache manually as well.
2. The app's purpose is for tracking traditional walking/hunting/navigational routes and monitoring by Indigenous peoples or other organizations.
3. Both platforms with identical functionality.
4. I just updated from version 2.0.2 to 4.0.0-dev.6 so this is a multipart answer. With version 2 it was very easy from what I remember so probably a 9. Upgrading to version 4 has been more challenging so more like a 6 or 7? I obviously had some initial troubles understanding how to set up bulk downloading. The documentation in the actual package files is helpful so I appreciate that. Obviously more documentation on the pub.dev page would be great but I understand why it's not there yet. I think the example being so long and kind of single big widget with everything included makes it more confusing than it is. You've obviously built a very thorough widget with multiple checks and options (which is awesome) but for my case I already check the network connectivity and want to show the download progress in a different way so I had to chop up your example to make it work for me. However, this might be exactly what some others want so I don't mean to say it's not useful.
5. I'm not sure if it's a bug or I'm doing something wrong but after I bulk download tiles, if I go into Airplane mode or just turn off all network connections, I get really choppy map performance (using `StorageCachingTileProvider`). The tiles are there but zooming in and out is glitchy (iOS and Android real devices). I get this exception (which is under a header called: 
` ======== Exception caught by image resource service ================================================`
`The following SocketException was thrown resolving an image codec:
Failed host lookup: 'api.mapbox.com' (OS Error: No address associated with hostname, errno = 7)`
I'm not sure what's going on but wanted to mention it in case it's a bug. My only guess is that a network call is still being attempted?

Finally, the only other thing that tripped me up was that I didn't realize that you needed to create your own folder if you want to have different/multiple stores apart from the `mainStore`. I assumed that if the folder wasn't created that it would get created automatically. It's no problem to create it but I didn't see it mentioned anywhere so adding a note somewhere would probably help some other people in the future.

My apologies if I'm just doing stuff wrong and then complaining about it! This package is fantastic and I really appreciate it! I'm just trying to be helpful with my feedback.

**Reply by JaffaKetchup** [reply]
Thanks for your answers!

Don't worry (if you were) about ignoring the warning, it happens to the best of us and I certainly understand that I need to improve documentation on things like this!

About the example just being a big block of code, I plan to actually create a whole example package at some point sooner or later, which would make it much easier to understand and take only the necessary parts.

I provide the pre-download checks functionality because it provides information about the battery as well, and should further reduce the code needed for many developers.

Performance issues are unfortunately to be expected until I can implement my own `ImageProvider`. At the moment tile provider's (like mine) override the `getImage` synchronous function (synchronous because it returns an `ImageProvider`), meaning that all of my file handling and checks have to be done synchronously before loading the tile to the screen. Likewise to the example, I plan to implement a custom `ImageProvider` (which will be asynchronous) at some point - I've experimented before, however, and it was quite complicated for me, so it may or may not end up in v4.
The example has further performance issues because I built it badly and rushed with getting everything to sort of work: this has led to many variables in many states and the widget being re-`setState`d almost constantly. Again, the example rewrite ahead should fix these issues at the expense of a bigger example package.

That network call is sort of a 'half-bug', ideally it wouldn't be logged, but I'm not sure how to silence it. Besides, using the caching rules (can't remember what they're called right now, but they are a enumerable property of `StorageCachingTileProvider`) should allow you to select whether a network call is made in the first place.

However, your last paragraph about the `storeName` functionality also seems to be a 'half-bug'. New stores are automatically generated when bulk-downloading; but thinking about it now, not necessarily when browse-caching. I don't know if you knew, but you can use `createStore()` on a `MapCachingManager` instance to force the creation of a new cache. Anyway, I'll try to fix this before release.

Again, thanks for your answers! If you need anymore help, feel free to ask here or in a new issue/discussion.

**Reply by bmason8** [reply]
That's great news about an example package. I'm sure it'll be a lot of work so it's appreciated!

No problem regarding the network call logging (I didn't properly explain that I'm getting more than just that log output). This is no problem because it's actually just a log message:
`flutter: SocketException: Failed host lookup: 'api.mapbox.com' (OS Error: nodename nor servname provided, or not known, errno = 8)`.
However, I'm getting and actual error that looks similar but includes a bunch of other stuff: I'll post it below but will need to alter it to remove my access token just fyi): Sorry for the formatting I didn't know how to fix it lol.

```
======== Exception caught by image resource service ========
The following SocketException was thrown resolving an image codec:
Failed host lookup: 'api.mapbox.com' (OS Error: nodename nor servname provided, or not known, errno = 8)

When the exception was thrown, this was the stack: 
#0      _NativeSocket.startConnect (dart:io-patch/socket_patch.dart:677:35)
#1      _RawSocket.startConnect (dart:io-patch/socket_patch.dart:1680:26)
#2      RawSocket.startConnect (dart:io-patch/socket_patch.dart:27:23)
#3      RawSecureSocket.startConnect (dart:io/secure_socket.dart:237:22)
#4      SecureSocket.startConnect (dart:io/secure_socket.dart:60:28)
#5      _ConnectionTarget.connect (dart:_http/http_impl.dart:2437:24)
#6      _HttpClient._getConnection.connect (dart:_http/http_impl.dart:2805:12)
#7      _HttpClient._getConnection (dart:_http/http_impl.dart:2810:12)
#8      _HttpClient._openUrl (dart:_http/http_impl.dart:2694:12)
#9      _HttpClient.getUrl (dart:_http/http_impl.dart:2573:48)
#10     NetworkToFileImage._downloadFromTheNetworkAndSaveToTheLocalFile (package:network_to_file_image/src/network_to_file_image.dart:204:57)
#11     NetworkToFileImage._loadAsync (package:network_to_file_image/src/network_to_file_image.dart:164:23)
#12     NetworkToFileImage.load (package:network_to_file_image/src/network_to_file_image.dart:120:16)
#13     ImageProvider.resolveStreamForKey.<anonymous closure> (package:flutter/src/painting/image_provider.dart:505:13)
#14     ImageCache.putIfAbsent (package:flutter/src/painting/image_cache.dart:380:22)
#15     ImageProvider.resolveStreamForKey (package:flutter/src/painting/image_provider.dart:503:83)
#16     ImageProvider.resolve.<anonymous closure> (package:flutter/src/painting/image_provider.dart:333:9)
#17     ImageProvider._createErrorHandlerAndKey.<anonymous closure>.<anonymous closure> (package:flutter/src/painting/image_provider.dart:464:26)
#18     SynchronousFuture.then (package:flutter/src/foundation/synchronous_future.dart:41:35)
#19     ImageProvider._createErrorHandlerAndKey.<anonymous closure> (package:flutter/src/painting/image_provider.dart:461:11)
#23     ImageProvider._createErrorHandlerAndKey (package:flutter/src/painting/image_provider.dart:453:16)
#24     ImageProvider.resolve (package:flutter/src/painting/image_provider.dart:330:5)
#25     Tile.loadTileImage (package:flutter_map/src/layer/tile_layer.dart:1267:36)
#26     _TileLayerState._addTile (package:flutter_map/src/layer/tile_layer.dart:1072:10)
#27     _TileLayerState._update (package:flutter_map/src/layer/tile_layer.dart:1014:7)
#35     throttleStreamTransformerWithTrailingCall.<anonymous closure> (package:flutter_map/src/core/util.dart:46:12)
#36     throttleStreamTransformerWithTrailingCall.<anonymous closure>.<anonymous closure> (package:flutter_map/src/core/util.dart:52:26)
(elided 21 frames from class _RawReceivePortImpl, class _Timer, dart:async, and dart:async-patch)
Image provider: NetworkToFileImage("/var/mobile/Containers/Data/Application/E669647B-8BC4-46D8-86B7-5738C87DD91B/Documents/mapCache/outdoors/apimapboxcomstylesv1mapboxoutdoors-v11tiles1425745667?access_token=MY_TOKEN", "https://api.mapbox.com/styles/v1/mapbox/outdoors-v11/tiles/14/2574/5667?access_token=MY_TOKEN", scale: 1.0)
File: /var/mobile/Containers/Data/Application/E669647B-8BC4-46D8-86B7-5738C87DD91B/Documents/mapCache/outdoors/apimapboxcomstylesv1mapboxoutdoors-v11tiles1425745667?access_token=MY_TOKEN
Url: https://api.mapbox.com/styles/v1/mapbox/outdoors-v11/tiles/14/2574/5667?access_token=MY_TOKEN
```
- this is @JaffaKetchup fixing this formatting -

I've done some more testing and either something more is going on or I'm just doing things incorrectly but this is what I've found:

1. If I use the `StorageCachingTileProvider` with only a `parentDirectory` it works great (same performance as I'm used to on v_2.. versions). Doesn't matter if I'm online or offline/airplane mode, it works great.
2. If I add the `storeName` parameter it causes the performance issues I mentioned before. I thought it was only if I was offline but it actually ruins performance when online as well. I've also tried with `behavior: CacheBehavior.cacheOnly` but it made no difference.

So am I doing something wrong with how I'm using the storeName parameter or could it be some type of bug? I didn't know you could use createStore() to create a new cache or that they are automatically generated when bulk downloading. I'm hoping to use the stores and bulk downloading in the following way:
- User has 2 map styles available to them. Anytime they switch styles, I switch the storeName so it caches the different styles (even for the same area) when panning around.
- If they enable 'Offline Maps' -> it requires them to choose which map style to have downloaded (ie storeName sub folder) and then downloads the tiles into that storeName subfolder. Are there any issues you see with that approach? Would duplicate tiles that are already in that store/sub folder just get overwritten or would that cause an issue?

Below is a bit of an example of how I'm using it.

This runs in a setup function before the map gets loaded.
`mapTilesParentDirectory = await MapCachingManager.normalCache;
    if (!mapTilesParentDirectory.existsSync()) {
      mapTilesParentDirectory.create();
      print('mapTilesParentDirectory created');
    }`

Then I use the tile provider like this:
```
TileProvider tileProvider;
    // To avoid a blank map -> use a NonCachingNetworkTileProvider until the setup is done (required because 'mapTilesParentDirectory' needs to be asynchronously fetched or created in setup)
    if (setupInProgress) {
      tileProvider = NonCachingNetworkTileProvider();
      print('Using NonCachingNetworkTileProvider()');
    } else {
      print('Using StorageCachingTileProvider(): storeName: $storeName');
      tileProvider = StorageCachingTileProvider(
        cachedValidDuration: Duration(days: 60),
        parentDirectory: mapTilesParentDirectory,
        storeName: storeName,
        // behavior: CacheBehavior.cacheOnly,
      );
    }
```

```
return TileLayerOptions(
        tileProvider: tileProvider,
        urlTemplate: "https://api.mapbox.com/styles/v1/mapbox/$type/tiles/{z}/{x}/{y}?access_token={accessToken}",
        additionalOptions: {'accessToken': mapboxApiToken},
        backgroundColor: Colors.blueGrey);
```

I also had the thought that maybe there was an issue because I wasn't creating the subfolder for the storeName so I added some code to create the folder at the end of the parentDirectory as well but it made no difference to performance either.
Thanks for your help and hopefully I'm just doing something stupid!

**Reply by bmason8** [reply]
After testing some more, I'm now seeing the same choppy map performance whenever (and only when) after bulk downloading tiles. Even if I never use a storeName when downloading and just use a parentDirectory. So if I just let the `StorageCachingTileProvider` cache on it's own, then go offline, it's great.
So I'm thinking I must be doing something wrong when downloading tiles.

**Reply by JaffaKetchup** [reply]
Hi again,

I fixed your formatting, so don't worry about that 😂. That error does seem to be because of a combination of my library, my dependencies, and `flutter_map`. I could potentially wrap the calls in `try{} catch{}` to silence that potential error: that error can be ignored if using this library safely, and I can't really prevent it easily.
If possible, could you try the same code and specify `CacheBehavior.cacheOnly` and see if the error still appears when offline?

However, this path from the log really worries me:
> /var/mobile/Containers/Data/Application/E669647B-8BC4-46D8-86B7-5738C87DD91B/Documents/mapCache/outdoors/apimapboxcomstylesv1mapboxoutdoors-v11tiles1425745667?access_token=MY_TOKEN

I can see my library's proper structure here: `<cache>/mapCache/<storeName>/<processedTileName>`, but that question mark character is an invalid path character on Android systems according to multiple sources. Have you tested your app on Android yet? If you can, please check if it works, if you can't, I'll check it on my side. If it does work, then that's a lucky miracle and needs to be urgently fixed.

---

> If I use the `StorageCachingTileProvider` with only a `parentDirectory` it works great (same performance as I'm used to on v_2.. versions). Doesn't matter if I'm online or offline/airplane mode, it works great.
If I add the `storeName` parameter it causes the performance issues I mentioned before. I thought it was only if I was offline but it actually ruins performance when online as well. I've also tried with `behavior: CacheBehavior.cacheOnly` but it made no difference.

This is extremely interesting to me, as I thought that there should theoretically be no difference, because omitting `storeName` just defaults to a store called 'mainStore'. If possible, could you create another store, maybe called 'secondStore' (or something else) and then delete the 'mainStore' (using the built-in `.deleteStore()` method), to see if the performance becomes ok again? If so, there may be performance problems when using multiple stores, potentially indicating a memory leak or running unnecessary code. I am sure you aren't doing anything wrong, this does seem to be my library.

---

> `mapTilesParentDirectory = await MapCachingManager.normalCache; if (!mapTilesParentDirectory.existsSync()) {mapTilesParentDirectory.create(); print('mapTilesParentDirectory created'); }`

I understand why you need this, but I'm not sure why I didn't spot this in testing. Hopefully my fix to auto create the store directories will also solve this recursively.

---

> User has 2 map styles available to them. Anytime they switch styles, I switch the storeName so it caches the different styles (even for the same area) when panning around.
If they enable 'Offline Maps' -> it requires them to choose which map style to have downloaded (ie storeName sub folder) and then downloads the tiles into that storeName subfolder. Are there any issues you see with that approach? Would duplicate tiles that are already in that store/sub folder just get overwritten or would that cause an issue?

This is a fairly common use-case, one I tried - and failed - to simplify back in v3 development. This should just work the way you described it. In answer to the first question, therefore, I can't see many issues except one: see #28. In answer to your second question, that depends on the `preventRedownload` setting when you create your `DownloadableRegion` from `.toDownloadable()`: if this is true, existing tiles (for example from browse caching) won't be overwritten, and vice-versa. Just make sure your internal variable and the `MapCachingManager` never go out of sync and start writing across stores; that would be quite messy.

Your code also looks good to me.

You appear to be doing everything right, so it's probably me 'at fault' :D

**Reply by JaffaKetchup** [reply]
Ooo, only just saw your latest response after I posted my long one!

I have also seen this performance issue in my example app.
Out of interest, what happens when you have only a hundred tiles to bulk download; ie. at what number of tiles does it start getting choppy?
Also, do you have any regular `setState`s in your app that might reload a `MapCachingManager().storeSize`?
One last question: what happens if you disable the recovery mode when bulk downloading? Does the lag go away?

Please make sure you only test as much as you can support whilst paying for your MapBox tiles (they aren't exactly cheap): if you can't test anymore, don't worry, just let me know :).

**Comment by bmason8** [comment]
I tried downloading tiles again with `preventRedownload` and it didn't make a difference.

I tried with `CacheBehavior.cacheOnly` and it made no difference unfortunately. I do get this message though while panning around with it on:
`LateInitializationError: Local 'bytes' has not been initialized.`

> I can see my library's proper structure here: <cache>/mapCache/<storeName>/<processedTileName>, but that question mark character is an invalid path character on Android systems according to multiple sources. Have you tested your app on Android yet? If you can, please check if it works, if you can't, I'll check it on my side. If it does work, then that's a lucky miracle and needs to be urgently fixed.
For all this last round of testing (I've using Android for this round of tests) after bulk downloading tiles, it does load them while offline (with jank) but only shows that error (that I posted the log of) when NOT setting `behavior: CacheBehavior.cacheOnly`. If `behavior: CacheBehavior.cacheOnly` is set, then I only get the harmless `SocketException: Failed host lookup: 'api.mapbox.com' (OS Error: No address associated with hostname, errno = 7)` message and as mentioned above `LateInitializationError: Local 'bytes' has not been initialized.`

Regarding the path that worries you with the '?'. Is that because you're just saving the tile with same name as the network call made to retrieve it? If so, that is probably not ideal if avoidable because that means the secret access token is stored in every single tile name. I'm not sure how the tiles need to be named for getting retrieved though so ignore that concern if it's required or not viewable by other apps.

> Out of interest, what happens when you have only a hundred tiles to bulk download; ie. at what number of tiles does it start getting choppy?
Also, do you have any regular setStates in your app that might reload a MapCachingManager().storeSize?
One last question: what happens if you disable the recovery mode when bulk downloading? Does the lag go away?
Please make sure you only test as much as you can support whilst paying for your MapBox tiles (they aren't exactly cheap): if you can't test anymore, don't worry, just let me know :).

Firstly, thanks for the reminder about downloading tiles! I totally wasn't thinking about the cost so I'm sure my company wouldn't want me racking up tile download costs! lol.
So during all my initial testing I was downloading about 10,000 tiles so I dropped that to about 1,000 and it made a noticeable difference. Still not as smooth as when there are no tiles downloaded via bulk download but smoother for sure. It's very interesting to me that after I delete any tiles that were bulk downloaded, the map is instantly back to being fully responsive.
`CacheDirectory directory = await MapCachingManager.normalCache;
    MapCachingManager(directory).deleteStore();`

Maybe it's a bug or maybe I'm downloading things incorrectly. I actually download each area at the same time so I'm going to do another round of tests where I only download a single large-ish area to see if that makes a difference.

> This is extremely interesting to me, as I thought that there should theoretically be no difference, because omitting storeName just defaults to a store called 'mainStore'. If possible, could you create another store, maybe called 'secondStore' (or something else) and then delete the 'mainStore' (using the built-in .deleteStore() method), to see if the performance becomes ok again? If so, there may be performance problems when using multiple stores, potentially indicating a memory leak or running unnecessary code. I am sure you aren't doing anything wrong, this does seem to be my library.

I haven't tried this but in the last two rounds of testing I haven't been using a store at all because the issue is present regardless. Seems like downloading tiles in bulk is messing things up (at least while saving them to the same main `mainStore` (ie not passing in a value and letting the default get used).
Let me know if there's anything more I can test!

**Reply by JaffaKetchup** [reply]
Well that's lucky that the file name still works on Android! Regarding your concern about whether having the API key in the filename is a security risk, it shouldn't be unless the device is rooted/jailbroken/emulated-with-adb because the storage space for the app is only accessible for that app.
However, my way of making valid and retrievable filenames isn't ideal at the moment, so to hit two birds with one stone, I'll hopefully use a quick hashing algorithm to remove invalid characters and hide the API keys a little better.

<details>
    <summary> Rough explanation of how I retrieve tiles if you're interested </summary>

When using the SQL database before v4, tiles were stored in a database, with their data and coordinates in separate columns. This avoided the need to use the file system, but meant that every retrieval had to look through the database to find the tile, taking time.
Since v4 (prior to above), the tile is stored almost exactly as the URL used to request it, filtering some banned characters. This meant that retrieval just had to have the URL and remove the banned characters, and it would have the tile.
There were also other reasons to move to a file based system.
</details>

I'll need to do some investigation to find the source of the `LateInitializationError`.

I thought that the cache only behaviour would silence the error and the log, as they seemed to be generated together, but apparently that's not the case. Again, I'll have to see if I can silence those, or better, fix those.

I kind of expected the performance to increase when there were less tiles, as I found clues to this in my example app. Unfortunately, I'm not sure why this is, but it is also an issue that needs to be fixed before release. The only reasons I can think of are memory leaks in my side, to much rebuilding in your side, or a limitation with flutter_map somehow (maybe similar to the reduced performance after many markers and polygons somehow idk).

Huge thanks for your help, it's made me realise some of my issues that I should've found myself (I need to test more). Hopefully I shouldn't need any more tests from you, but if you do find anything interesting, I'll be here!

I'll also keep you updated on this thread with the fixes mentioned above if you want?

**Reply by bmason8** [reply]
Just tried downloading only 1 area of about 3000 tiles via bulk downloading and the map performance issues remained.

I also panned around to add some more tiles to the cache automatically which brought the tile count up to around 3700 at 3.52 MB and the jank remained. On v_2.0... I could download over 50 MB and never saw any performance issues (just for comparison).

I also got curious and wondered if it's just a number of tiles limit issue rather than a bulk download issue so I ran `MapCachingManager(directory).deleteStore();` and just started panning around to the point where I got to 2 MB of tiles and sure enough the jank started to show itself. So it seems like something is going on and it only takes a couple of MB before it starts to show up. Hope this is helpful!

**Reply by bmason8** [reply]
Woops, I didn't see your comment come in and just realized I replied to it (thought I was just replying to my last comment).

I don't think I'm doing any excessive rebuilding since it happens even when just panning around the map with nothing else going on and only a couple of other map layers (polygons). Even if I remove the polygons the jank remains too.
Maybe it has something to do with Flutter Map. I only just upgraded it to v_0.14 (from v_0.13) so I could update this package so I can't comment on if I've noticed any performance hits.

Updates on any of this would be great if you could tag me or however it works.
Thanks for the explanations in your last comment, I'm not worried about the token in the name knowing that but maybe it's causing some issues due to the types of characters and would be good to alter regardless.

I appreciate all your work on this and quick replies! I'm sure we'll chat some more about this.
Cheers.

**Reply by JaffaKetchup** [reply]
Don't worry about backdating `flutter_map`, can't see why that would be it.

One more interesting point: I'm not sure the size of the store is accurate, I think my size-checking algorithm needs some work.

Anyway, likewise to you: thanks for your help!
Enjoy your weekend (soon?) :)

**Reply by JaffaKetchup** [reply]
Hi again @bmason8,
I've just released v4.0.0-dev.7 on pub.dev. Would you please update to it and test it out to see if it has fixed the issues we've discussed (as below)?

- Fixed performance issues through custom `ImageProvider`
- Better URL > filename processing
- Removal of `LateInitializationErrors`
- Made internal constructors harder to use by mistake (made them named constructors)
- Made `StorageCachingTileProvider` and `MapCachingManager` automatically create the cache store on initialization
- Added method to `MapCachingManager` to watch for statistic changes through filesystem changes

I haven't yet fixed or changed:
 - The example
 - Statistic getter inaccuracies? (more testing needed on my side)

Hope it works for you!

**Reply by bmason8** [reply]
Just updated and gave it a quick initial round of testing on both iOS and Android and it's working great for the most part! I only found one bug so far. Here's a quick summary of what I've found:

1. Performance issues are fixed!
2. No `LateInitializationErrors`
3. Parent directory was created automatically just by running `await MapCachingManager.normalCache`.
4. StoreName cache was created automatically by just passing a value into the constructor.

The one bug I found relates to swapping between storeName caches. I'll explain the steps I took to cause the issue below but basically when swapping back and forth between store caches it seems to just show whatever style (cache) was already downloaded for that part of the map.

So I did a fresh install and included a `storeName` called **outdoors** in the `StorageCachingTileProvider` constructor so it wouldn't use the default (`mainStore`). So right away anything it caches is getting stored there (in **outdoors**) which is perfect.

But when I press my 'mapSwap' button which:
-  changes which style of tile I request from mapbox
- updates the `storeName` variable used in the `StorageCachingTileProvider` to **satellite** (or back to **outdoors**)

The map doesn't actually load in the new **satellite** style tiles in areas where there are already **outdoors** style tiles. Instead it loads in the **outdoors** tiles and only when zooming in or out or panning around to uncached areas will it load in (and cache) the correct **satellite** tiles. So nothing is broken (ie tiles always get loaded) but the expected behaviour was that if I changed the `storeName` variable, it would only pull from that folder.

So when _online_, if I was over an area that was cached to **outdoors** and swapped to **satellite**, it would swap folders and only check for cached **satellite** tiles. And if they didn't exist, cache them like normal.

And if _offline_, (in the same scenario) it would just not show any tiles when trying to show **satellite** because those tiles weren't cached yet.

Let me know if I can provide any more details! This version is awesome though!

**Reply by JaffaKetchup** [reply]
That's great to hear!

To fix most of the issues in the end just needed some 'targeted tweaking' and I solved the performance issues by implementing my own `ImageProvider`.
Interestingly, a custom `ImageProvider` was used in v2.0.2, which probably explained why the performance was so much better. The whole point of one is to perform all of the caching asynchronously: when you started this discussion, everything was done synchronously on the same thread before showing the tile, hence the extreme lag.

To solve your last issue (I might be wrong here, please correct me if I am), please see https://github.com/JaffaKetchup/flutter_map_tile_caching/issues/28#issuecomment-949634839 & https://github.com/fleaflet/flutter_map/pull/667. What you're describing sounds like an issue with `flutter_map` that can be fixed by using the 'reset' functionality provided in the latest version. So every time you change the `storeName` also do what I linked above.

Hope that solves your issue - again, please correct me if I'm wrong; I'm always happy to help!

**Reply by bmason8** [reply]
Ah ok, good to know. I took at look at the pages you linked above and implemented the _resetTiles_ functionality demonstrated in the example lib but it's not working for me yet. I suspect it's something to do with how I'm doing it though (or a flutter_map/mapbox combo issue) so I'll keep working on it and hopefully figure it out. I think I read somewhere that it was working for them so I'm fairly certain it's an issue on my end. Thanks for all your help on this!

**Reply by JaffaKetchup** [reply]
Interesting. I don't think it's your combo though (flutter_map should work with any WMS raster source (which MapBox (usually) is)).
I'll keep testing on my end to see if I can resolve it, or whether it is flutter_map's framework causing the problem.

**Reply by bmason8** [reply]
Ok great, thanks! Let me know if you find anything. If not, I'll assume it's something I need to fix on my end.

**Reply by bmason8** [reply]
@JaffaKetchup I've yet to figure out how to swap back and forth between storeName's/tile sets without the caching issue where the previous tile type remains at certain zoom levels (often several). Like I mentioned before, I've tried the _resetTiles_ strategy from `flutter_map` [as shown here](https://github.com/fleaflet/flutter_map/blob/master/example/lib/pages/reset_tile_layer.dart) with all sorts of variations but nothing has worked.
The reason I'm mentioning this again is because the issue is fixed if I use `StorageCachingTileProvider` for one tileset and `NonCachingNetworkTileProvider` for the other. That seems odd to me so thought I'd let you know in case it gives you any ideas.

**Reply by JaffaKetchup** [reply]
In my testing so far in the new example app, I've only been using an OSM source. However, the reset thing seems to do the right thing for me (shows a blank map for a few frames) when switching between stores.

In a few days, I'll hope to have added custom sources to the example app, so I can do more testing then.

Sorry for the inconvenience :)!

**Reply by bmason8** [reply]
No worries, I appreciate your effort on this!
Ok great, let me know if you find any issues. I'll keep hacking away on my end and hopefully figure out what's going on.

**Reply by JaffaKetchup** [reply]
@bmason8 Hi,

I've been writing some more of the new example app, this time properly implementing custom sources per store, and I can still confirm that this is an issue.
I'm pretty sure it's not crossing stores (which is a good thing), so I think the issue is with the the way I've implemented my `ImageProvider` to solve those performance issues? I have to call `PaintingBinding.instance?.imageCache?.evict(this);` if the call for a tile fails (otherwise the tile will never load in, even if the problem resolves itself), and I have a feeling I need a similar thing to completely clear the image cache after a store switch.
Hopefully I can implement this fairly easily.

In other news, there's two new features:
 - `coverImage()` in `MapCachingManager`, generates a cover image for a store.
 - `MapMetadataManager`, a class 'interface' to persist map caching related data across restarts, separate to the main tile stores - for example to store the last used store or to store a source URL per cache store.

These will be pushed to GitHub, along with the work on the example app, soon.

**Reply by bmason8** [reply]
Awesome, that's good to hear. I tried so many ways to confirm what I was doing wasn't the issue but was still unsure so now I feel a little less crazy haha. Those additions sounds great as well.
Thanks again!

**Reply by JaffaKetchup** [reply]
Hi @bmason8,
Sorry, this is taking a really long time!

It turns out this is not going to be as easy fix as I thought, so for the time being, you may want to do one of these things:
 - Rollback to a previous version where this works for you
 - Find a workaround (perhaps using persistent storage to remember the last used store, like in the example, then using [flutter_phoenix](https://pub.dev/packages/flutter_phoenix))
 - Put up with the issue until I can fix it

I'm really sorry for the inconvenience.

**Reply by bmason8** [reply]
@JaffaKetchup No worries! I'm about to push out an update using the dev-7 version and will be using the `NonCachingNetworkTileProvider` for the satellite tiles until a fix is ready. My users are used to that behaviour anyways so it's no big deal. Thanks for all your work on this!

**Reply by JaffaKetchup** [reply]
@bmason8 Please see https://github.com/JaffaKetchup/flutter_map_tile_caching/pull/15#issuecomment-984856798 and #31. #31 will be the new issue for all further discussion on this topic.
Thanks!
